File: windows/DotsHarness/src/PluginRuntime/PluginCatalog.cs
// Copyright (c) 2026 DOTS
// Plugin composition model derived from DeepSeek Harness.
// Copyright (c) 2026 DeepSeek. MIT. See NOTICE.

using System.Text.Json;
using System.Text.Json.Serialization;
using HarnessPluginKit;

namespace PluginRuntime;

public enum PluginKind
{
    Builtin,
    Manifest,
    Dylib,
}

public sealed record CatalogEntry(
    PluginManifest Manifest,
    PluginKind Kind,
    PluginTrust Trust,
    bool Enabled = true,
    string? Broken = null,
    string? Url = null)
{
    public string Id => Manifest.Id;
}

public sealed record ResolvedPlugin(
    PluginManifest Manifest,
    PluginKind Kind,
    PluginTrust Trust,
    Func<IHarnessPlugin> Make);

public sealed record SupportPaths(
    string Root,
    string Plugins,
    string Presets,
    string Settings,
    string HostPatch,
    string Trust,
    string Models,
    string Runtime)
{
    public static SupportPaths Default()
    {
        var baseDir = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        if (string.IsNullOrEmpty(baseDir))
        {
            baseDir = Path.GetTempPath();
        }
        var root = Path.Combine(baseDir, "DotsHarness");
        return new SupportPaths(
            root,
            Path.Combine(root, "plugins"),
            Path.Combine(root, "presets"),
            Path.Combine(root, "settings.json"),
            Path.Combine(root, "host.patch.yml"),
            Path.Combine(root, "trust.json"),
            Path.Combine(root, "models"),
            Path.Combine(root, "runtime"));
    }

    public void Ensure()
    {
        Directory.CreateDirectory(Plugins);
        Directory.CreateDirectory(Presets);
        Directory.CreateDirectory(Models);
        Directory.CreateDirectory(Runtime);
    }
}

public sealed class PluginCatalog : ObservableObject
{
    private readonly Dictionary<string, Func<IHarnessPlugin>> _builtins = new(StringComparer.Ordinal);
    private readonly Dictionary<string, PluginManifest> _builtinManifests = new(StringComparer.Ordinal);
    private readonly Dictionary<string, PluginTrust> _trustOverrides = new(StringComparer.Ordinal);
    private readonly HashSet<string> _disabled = new(StringComparer.Ordinal);
    private IReadOnlyList<CatalogEntry> _entries = Array.Empty<CatalogEntry>();

    public SupportPaths Paths { get; }

    public IReadOnlyList<CatalogEntry> Entries
    {
        get => _entries;
        private set => SetProperty(ref _entries, value);
    }

    public PluginCatalog(SupportPaths? paths = null)
    {
        Paths = paths ?? SupportPaths.Default();
        Paths.Ensure();
        LoadTrust();
    }

    public void RegisterBuiltin(Func<IHarnessPlugin> factory)
    {
        var plugin = factory();
        var manifest = plugin.Manifest;
        _builtinManifests[manifest.Id] = manifest;
        _builtins[manifest.Id] = factory;
        Refresh();
    }

    public void RegisterBuiltin<T>() where T : IDefaultPlugin, new() => RegisterBuiltin(() => new T());

    public void Refresh()
    {
        var next = new List<CatalogEntry>();
        foreach (var (id, manifest) in _builtinManifests.OrderBy(kv => kv.Key, StringComparer.Ordinal))
        {
            next.Add(new CatalogEntry(manifest, PluginKind.Builtin, PluginTrust.System, !_disabled.Contains(id)));
        }

        if (Directory.Exists(Paths.Plugins))
        {
            foreach (var folder in Directory.GetDirectories(Paths.Plugins))
            {
                var yaml = Path.Combine(folder, "plugin.yml");
                if (!File.Exists(yaml)) continue;
                try
                {
                    var manifest = MiniYaml.DecodeManifest(File.ReadAllText(yaml));
                    var kind = manifest.Library is null ? PluginKind.Manifest : PluginKind.Dylib;
                    var trust = _trustOverrides.TryGetValue(manifest.Id, out var t) ? t : PluginTrust.Untrusted;
                    next.Add(new CatalogEntry(manifest, kind, trust, !_disabled.Contains(manifest.Id), Url: folder));
                }
                catch (Exception ex)
                {
                    var name = Path.GetFileName(folder);
                    next.Add(new CatalogEntry(
                        new PluginManifest(name, name, "0.0.0", PluginPlane.Session),
                        PluginKind.Manifest,
                        PluginTrust.Untrusted,
                        Enabled: false,
                        Broken: ex.Message,
                        Url: folder));
                }
            }
        }

        Entries = next;
    }

    public void SetTrust(string id, PluginTrust trust)
    {
        if (_builtinManifests.ContainsKey(id)) return;
        _trustOverrides[id] = trust;
        PersistTrust();
        Refresh();
    }

    public void SetEnabled(string id, bool isEnabled)
    {
        if (isEnabled) _disabled.Remove(id);
        else _disabled.Add(id);
        PersistTrust();
        Refresh();
    }

    public ResolvedPlugin Resolve(string id)
    {
        if (_builtins.TryGetValue(id, out var factory) && _builtinManifests.TryGetValue(id, out var builtin))
        {
            return new ResolvedPlugin(builtin, PluginKind.Builtin, PluginTrust.System, factory);
        }

        var entry = Entries.FirstOrDefault(e => e.Manifest.Id == id)
            ?? throw PluginException.UnknownPlugin(id);
        if (entry.Broken is { } broken) throw PluginException.InvalidManifest(broken);

        return entry.Kind switch
        {
            PluginKind.Builtin => throw PluginException.UnknownPlugin(id),
            PluginKind.Manifest => new ResolvedPlugin(entry.Manifest, PluginKind.Manifest, entry.Trust,
                () => new ManifestPlugin(entry.Manifest, entry.Url)),
            PluginKind.Dylib => ResolveLibrary(entry),
            _ => throw PluginException.UnknownPlugin(id),
        };
    }

    private static ResolvedPlugin ResolveLibrary(CatalogEntry entry)
    {
        if (entry.Url is null || entry.Manifest.Library is null)
        {
            throw PluginException.InvalidManifest("missing library");
        }
        var path = Path.Combine(entry.Url, entry.Manifest.Library);
        return new ResolvedPlugin(entry.Manifest, PluginKind.Dylib, entry.Trust, () => PluginLibrary.Load(path));
    }

    private void LoadTrust()
    {
        if (!File.Exists(Paths.Trust)) return;
        try
        {
            var file = JsonSerializer.Deserialize<TrustFile>(File.ReadAllText(Paths.Trust), TrustJson.Options);
            if (file is null) return;
            _trustOverrides.Clear();
            foreach (var (key, value) in file.Trust)
            {
                _trustOverrides[key] = value;
            }
            _disabled.Clear();
            foreach (var id in file.Disabled) _disabled.Add(id);
        }
        catch
        {
            // ignore corrupt trust file
        }
    }

    private void PersistTrust()
    {
        var file = new TrustFile(
            new Dictionary<string, PluginTrust>(_trustOverrides, StringComparer.Ordinal),
            _disabled.OrderBy(x => x, StringComparer.Ordinal).ToList());
        File.WriteAllText(Paths.Trust, JsonSerializer.Serialize(file, TrustJson.Options));
    }

    private sealed record TrustFile(Dictionary<string, PluginTrust> Trust, List<string> Disabled);

    private static class TrustJson
    {
        public static readonly JsonSerializerOptions Options = new()
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) },
            WriteIndented = true,
        };
    }
}

File: macos/DotsHarness/Sources/PluginRuntime/PluginCatalog.swift
// Copyright (c) 2026 DOTS
// Plugin catalog and support paths for Dots Harness.

import Foundation
import HarnessPluginKit

public enum PluginKind: String, Sendable {
    case builtin
    case dylib
    case unsupported
}

public struct CatalogEntry: Identifiable, Sendable {
    public var id: String { manifest.id }
    public var manifest: PluginManifest
    public var kind: PluginKind
    public var trust: PluginTrust
    public var enabled: Bool
    public var broken: String?
    public var url: URL?

    public init(
        manifest: PluginManifest,
        kind: PluginKind,
        trust: PluginTrust,
        enabled: Bool = true,
        broken: String? = nil,
        url: URL? = nil
    ) {
        self.manifest = manifest
        self.kind = kind
        self.trust = trust
        self.enabled = enabled
        self.broken = broken
        self.url = url
    }
}

public struct ResolvedPlugin {
    public var manifest: PluginManifest
    public var kind: PluginKind
    public var trust: PluginTrust
    public var directory: URL?
    public var make: () throws -> HarnessPlugin

    public init(
        manifest: PluginManifest,
        kind: PluginKind,
        trust: PluginTrust,
        directory: URL? = nil,
        make: @escaping () throws -> HarnessPlugin
    ) {
        self.manifest = manifest
        self.kind = kind
        self.trust = trust
        self.directory = directory
        self.make = make
    }
}

public struct SupportPaths: Sendable {
    public var root: URL
    public var plugins: URL
    public var presets: URL
    public var settings: URL
    public var hostPatch: URL
    public var trust: URL
    public var models: URL
    public var runtime: URL

    public init(
        root: URL,
        plugins: URL,
        presets: URL,
        settings: URL,
        hostPatch: URL,
        trust: URL,
        models: URL,
        runtime: URL
    ) {
        self.root = root
        self.plugins = plugins
        self.presets = presets
        self.settings = settings
        self.hostPatch = hostPatch
        self.trust = trust
        self.models = models
        self.runtime = runtime
    }

    public static func `default`(fileManager: FileManager = .default) -> SupportPaths {
        let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        let root = base.appendingPathComponent("DotsHarness", isDirectory: true)
        return SupportPaths(
            root: root,
            plugins: root.appendingPathComponent("plugins", isDirectory: true),
            presets: root.appendingPathComponent("presets", isDirectory: true),
            settings: root.appendingPathComponent("settings.json"),
            hostPatch: root.appendingPathComponent("host.patch.yml"),
            trust: root.appendingPathComponent("trust.json"),
            models: root.appendingPathComponent("models", isDirectory: true),
            runtime: root.appendingPathComponent("runtime", isDirectory: true)
        )
    }

    /// Pinned publisher public keys for signed marketplace packages.
    public var publishers: URL { root.appendingPathComponent("publishers.json") }

    public func ensure(fileManager: FileManager = .default) {
        try? fileManager.createDirectory(at: plugins, withIntermediateDirectories: true)
        try? fileManager.createDirectory(at: presets, withIntermediateDirectories: true)
        try? fileManager.createDirectory(at: models, withIntermediateDirectories: true)
        try? fileManager.createDirectory(at: runtime, withIntermediateDirectories: true)
        try? fileManager.createDirectory(
            at: root.appendingPathComponent("skills", isDirectory: true),
            withIntermediateDirectories: true
        )
    }
}

public final class PluginCatalog: ObservableObject {
    @Published public private(set) var entries: [CatalogEntry] = []

    public let paths: SupportPaths
    private var builtins: [String: () -> HarnessPlugin] = [:]
    private var builtinManifests: [String: PluginManifest] = [:]
    private var trustOverrides: [String: PluginTrust] = [:]
    private var disabled: Set<String> = []
    private var didLoadTrust = false

    public init(paths: SupportPaths = .default()) {
        self.paths = paths
    }

    public func registerBuiltin(_ plugin: HarnessPlugin.Type, refresh: Bool = true) {
        let manifest = plugin.manifest
        builtinManifests[manifest.id] = manifest
        builtins[manifest.id] = { plugin.initForCatalog() }
        if !entries.contains(where: { $0.manifest.id == manifest.id }) {
            entries.append(CatalogEntry(
                manifest: manifest,
                kind: .builtin,
                trust: .system,
                enabled: !disabled.contains(manifest.id)
            ))
        }
        if refresh {
            self.refresh()
        }
    }

    public func refresh() {
        paths.ensure()
        if !didLoadTrust {
            loadTrust()
            didLoadTrust = true
        }

        var next: [CatalogEntry] = []
        for (id, manifest) in builtinManifests.sorted(by: { $0.key < $1.key }) {
            next.append(CatalogEntry(
                manifest: manifest,
                kind: .builtin,
                trust: .system,
                enabled: !disabled.contains(id)
            ))
        }
        let fm = FileManager.default
        if let children = try? fm.contentsOfDirectory(at: paths.plugins, includingPropertiesForKeys: nil) {
            for folder in children where folder.hasDirectoryPath {
                let yaml = folder.appendingPathComponent("plugin.yml")
                guard fm.fileExists(atPath: yaml.path) else { continue }
                do {
                    let text = try String(contentsOf: yaml, encoding: .utf8)
                    let manifest = try MiniYAML.decode(PluginManifest.self, from: text)
                    let trust = trustOverrides[manifest.id] ?? .untrusted
                    let runtime = manifest.runtime.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                    let library = manifest.library?.trimmingCharacters(in: .whitespacesAndNewlines)
                    guard runtime == "native", let library, !library.isEmpty,
                          !library.hasPrefix("/"), !library.contains(".."),
                          !library.contains("\\"),
                          fm.fileExists(atPath: folder.appendingPathComponent(library).path) else {
                        next.append(CatalogEntry(
                            manifest: manifest,
                            kind: .unsupported,
                            trust: trust,
                            enabled: false,
                            broken: "Only native plugins with a safe `library` are supported. JavaScript and declarative plugins are no longer loadable.",
                            url: folder
                        ))
                        continue
                    }
                    next.append(CatalogEntry(
                        manifest: manifest,
                        kind: .dylib,
                        trust: trust,
                        enabled: !disabled.contains(manifest.id),
                        url: folder
                    ))
                } catch {
                    next.append(CatalogEntry(
                        manifest: PluginManifest(
                            id: folder.lastPathComponent,
                            name: folder.lastPathComponent,
                            version: "0.0.0",
                            plane: .session
                        ),
                        kind: .unsupported,
                        trust: .untrusted,
                        enabled: false,
                        broken: error.localizedDescription,
                        url: folder
                    ))
                }
            }
        }
        entries = next
    }

    public func setTrust(_ id: String, _ trust: PluginTrust) {
        guard builtinManifests[id] == nil else { return }
        trustOverrides[id] = trust
        persistTrust()
        refresh()
    }

    public func setEnabled(_ id: String, _ isEnabled: Bool) {
        if isEnabled { disabled.remove(id) } else { disabled.insert(id) }
        persistTrust()
        refresh()
    }

    public func resolve(_ id: String) throws -> ResolvedPlugin {
        if let factory = builtins[id], let manifest = builtinManifests[id] {
            return ResolvedPlugin(manifest: manifest, kind: .builtin, trust: .system, make: factory)
        }
        guard let entry = entries.first(where: { $0.manifest.id == id }) else {
            throw PluginError.unknownPlugin(id)
        }
        if let broken = entry.broken {
            throw PluginError.invalidManifest(broken)
        }
        switch entry.kind {
        case .builtin:
            throw PluginError.unknownPlugin(id)
        case .unsupported:
            throw PluginError.invalidManifest("Only native plugins with a compiled library can be loaded")
        case .dylib:
            guard let folder = entry.url, let library = entry.manifest.library else {
                throw PluginError.invalidManifest("missing library")
            }
            let url = folder.appendingPathComponent(library)
            return ResolvedPlugin(manifest: entry.manifest, kind: .dylib, trust: entry.trust, directory: folder) {
                try PluginDylib.load(from: url)
            }
        }
    }

    private func loadTrust() {
        guard let data = try? Data(contentsOf: paths.trust),
              let object = try? JSONDecoder().decode(TrustFile.self, from: data) else { return }
        trustOverrides = object.trust
        disabled = Set(object.disabled)
    }

    private func persistTrust() {
        let file = TrustFile(trust: trustOverrides, disabled: Array(disabled).sorted())
        if let data = try? JSONEncoder().encode(file) {
            try? data.write(to: paths.trust, options: .atomic)
        }
    }
}

private struct TrustFile: Codable {
    var trust: [String: PluginTrust]
    var disabled: [String]
}

private extension HarnessPlugin {
    static func initForCatalog() -> HarnessPlugin {
        if let type = self as? any DefaultPlugin.Type {
            return type.create()
        }
        fatalError("Built-in plugin \(manifest.id) must conform to DefaultPlugin")
    }
}
